examples/argobots/sendrecv_ult.c

```diff
@@ -0,0 +1,149 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ * See COPYRIGHT in top-level directory.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <abt.h>
+#include "mpi.h"
+
+#define DEFAULT_NUM_THREADS     2
+#define NUM_LOOP                10
+#define BUF_SIZE                16384
+
+#define HANDLE_ERROR(ret,msg)                           \
+    if (ret != ABT_SUCCESS) {                           \
+        fprintf(stderr, "ERROR[%d]: %s\n", ret, msg);   \
+        exit(EXIT_FAILURE);                             \
+    }
+
+int rank;
+int size;
+int buf_size;
+int *send_buf;
+int *recv_buf;
+int num_loop;
+int sum;
+int verbose = 0;
+int profiling = 1;
+double t1, t2;
+#define FIELD_WIDTH 16
+
+void thread_send_func(void *arg)
+{
+    ABT_thread ult_self__;
+    ABT_thread_self(&ult_self__);
+    size_t my_id = (size_t) arg;
+    if (verbose)
+        printf("[rank %d, TH%lu]: send\n", rank, my_id);
+
+    int i, j;
+    for (i = 0; i < num_loop; i++) {
+        for (j = 1; j < size; j++) {
+            int dest = (rank + j) % size;
+            send_buf[0] = rank;
+            if (verbose)
+                printf("[rank %d, TH%lu]: MPI_Send to %d\n", rank, my_id,
+                       dest);
+            MPI_Send(send_buf, buf_size, MPI_INT, dest, 0, MPI_COMM_WORLD);
+        }
+    }
+
+}
+
+void thread_recv_func(void *arg)
+{
+    ABT_thread ult_self__;
+    ABT_thread_self(&ult_self__);
+    size_t my_id = (size_t) arg;
+    if (verbose)
+        printf("[rank %d, TH%lu]: receive\n", rank, my_id);
+
+    int i, j;
+    for (i = 0; i < num_loop; i++) {
+        for (j = 1; j < size; j++) {
+            int source = (rank + size - j) % size;
+            if (verbose)
+                printf("[rank %d, TH%lu]: MPI_Recv from %d\n", rank, my_id,
+                       source);
+            MPI_Recv(recv_buf, buf_size, MPI_INT, source, 0,
+                     MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+            sum += recv_buf[0];
+        }
+    }
+}
+
+int main(int argc, char *argv[])
+{
+    int i, provided;
+    int ret;
+    int num_threads = DEFAULT_NUM_THREADS;
+    if (argc > 1)
+        num_threads = atoi(argv[1]);
+    if (num_threads % 2 != 0) {
+        if (rank == 0)
+            printf("number of user level threads should be even\n");
+        exit(0);
+    }
+    assert(num_threads >= 0);
+    num_loop = NUM_LOOP;
+    if (argc > 2)
+        num_loop = atoi(argv[2]);
+    assert(num_loop >= 0);
+    buf_size = BUF_SIZE;
+    if (argc > 3)
+        buf_size = atoi(argv[3]);
+    assert(buf_size >= 0);
+    send_buf = (int *) malloc(buf_size * sizeof(int));
+    recv_buf = (int *) malloc(buf_size * sizeof(int));
+
```

Comment:
IMO, the user should initialize Argobots before MPI. See also below

Reply:
👍